platforms/core-runtime/gradle-cli/src/main/java/org/gradle/launcher/cli/DefaultCommandLineActionFactory.java

```diff
@@ -166,25 +181,37 @@ private static class CommandLineParseFailureAction implements Action<ExecutionLi
         private final Exception exception;
         private final CommandLineParser parser;
 
-        public CommandLineParseFailureAction(CommandLineParser parser, Exception exception) {
+        private final List<String> args;
+
+        public CommandLineParseFailureAction(CommandLineParser parser, Exception exception, List<String> args) {
             this.parser = parser;
             this.exception = exception;
+            this.args = args;
         }
 
         @Override
         public void execute(ExecutionListener executionListener) {
             System.err.println();
             System.err.println(exception.getMessage());
-            showUsage(System.err, parser);
+            String suggestedTaskName = null;
+            for (String arg : args) {
+                if (!arg.startsWith("-")) {
+                    suggestedTaskName = arg;
+                    break;
+                }
+            }
+            showUsage(System.err, parser, suggestedTaskName);
```

Comment:
💭  One thing to consider is that this finds a _task selector_, not a task name per se. Maybe it's better to use that terminology to avoid unsupported assumptions.

Also, we should consider (and test) the UX of when it was indeed a task selector. What help would the user get if they run `gradle help --task codeQ`?

Reply:
Task selectors work with help task:
```
% gradle help --task ini
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :help
Detailed task information for ini

Path
     :init
```